### PR TITLE
Refactor scout engines docker image

### LINKED ISSUE #64: Change Scout images to use build-bench-base / llvm-10
This should only focus on changing `scout-engines/Dockerfile`  to use `ewasm/llvm-10:1` as a base and remove all the "workarounds" off installing cmake, llvm, etc.

### PR DESCRIPTION
**Pull request by jwasinger**
* Reduces final image size via multistage build
* Bases the image off `ewasm/bench-build-base`

Closes #64.

### CODE REVIEW

**Comment by jwasinger** [comment]
Will squash `RUN` statements after I verify that these changes work.

**Comment by cdetrio** [comment]
This deletes all the comments (some of them are useful).

**Comment by cdetrio** [comment]
Why does it install node 11?

**Comment by jwasinger** [comment]
> Why does it install node 11?

Good catch.  should be installing node 12 imo.

**Comment by jwasinger** [comment]
> This deletes all the comments (some of them are useful).

I'll add them back in

**Comment by cdetrio** [comment]
> Good catch. should be installing node 12 imo.

There was already a line to install node 12, btw.

**Comment by axic** [comment]
Make sure we compare the benchmarks with and without this PR before merging. If the `llvm-10` image has a different compiler version that may produce different results -- hopefully faster output.

**Comment by axic** [comment]
@jwasinger can you address the comments?

**Comment by jwasinger** [comment]
I've addressed the comments in the latest commit.

**Comment by jwasinger** [comment]
> Make sure we compare the benchmarks with and without this PR before merging. If the `llvm-10` image has a different compiler version that may produce different results -- hopefully faster output.

I'll check this

File: scout-engines/Dockerfile [hunk 1]
```diff
@@ -246,14 +199,8 @@ RUN cd eip1962-bls12-rs-native && cargo build --release
...
-# magic to install cmake 13.5 on ubuntu 19.10
-RUN wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | gpg --dearmor - | sudo tee /etc/apt/trusted.gpg.d/kitware.gpg >/dev/null && \
-  wget http://archive.ubuntu.com/ubuntu/pool/main/o/openssl1.0/libssl1.0.0_1.0.2n-1ubuntu5.3_amd64.deb && \
-  dpkg -i libssl1.0.0_1.0.2n-1ubuntu5.3_amd64.deb && \
-  apt-add-repository 'deb https://apt.kitware.com/ubuntu/ xenial main' && \
-  apt-get update && \
-  apt-get install -y cmake-data=3.15.5-0kitware1 && \
-  apt-get install -y cmake=3.15.5-0kitware1
+ENV DEBIAN_FRONTENT=noninteractive
+RUN apt install -y cmake
```
Comment (axic) [inline scout-engines/Dockerfile:203]: cmake is added by `bench-build-base` already

File: scout-engines/Dockerfile [hunk 2]
```diff
@@ -172,9 +121,13 @@ RUN cd wabt-bls12-bignums-fasthost-fastmont-no-superops && make clang-release -j
 #RUN git clone --recursive --single-branch --branch bignum-hostfuncs-bn128-rolluprs https://github.com/ewasm-benchmarking/scout_wabt.cpp.git scoutcpp-bn128-rolluprs
 #RUN cd scoutcpp-bn128-rolluprs && mkdir build && cd build && cmake -DBUILD_TESTS=OFF -DCMAKE_BUILD_TYPE=Release .. && make -j4
 
-
 WORKDIR /scoutyamls
 
+
+# install node 12
+RUN curl -sL https://deb.nodesource.com/setup_12.x | bash
```
Comment (axic) [inline scout-engines/Dockerfile:128]: @cdetrio this is actually moved from an earlier place in the docker image, not added as a new. Apparently it was always comparing against node12 in scout-engines.

File: scout-engines/Dockerfile [hunk 3]
```diff
@@ -1,53 +1,8 @@
...
-
-RUN apt-get install -y python # python2.7 needed to build wabt
-
-RUN apt-get install -y python3 python3-distutils
-RUN wget https://bootstrap.pypa.io/get-pip.py && python3 get-pip.py
-
-## pandas needed for nanodurationpy (for the benchmarking script)
-RUN pip3 install pandas
-
+# FROM ewasm/bench-build-base:1 as build
```
Comment (axic) [inline scout-engines/Dockerfile:1]: Stray comment.

**Comment by axic** [comment]
So despite changing the base image, both use llvm-10/clang-10 as the default compiler, and thus I don't think there should be any measurable speed difference with the generated binaries.

@cdetrio @jwasinger what is left to merge this? Did anyone compare pre/post times?

**Comment by jwasinger** [comment]
I've pushed the dataset and charts from this branch to https://github.com/ewasm/benchmarking/pull/101 (with the intention of having this branch merged and then merging 101).  There is no substantial difference in the charts that I looked at.

**Comment by jwasinger** [comment]
```
> ethereumjs-config-build

+ exec tsc -p ./tsconfig.prod.json
src/lib.ts(247,19): error TS2532: Object is possibly 'undefined'.
src/lib.ts(247,28): error TS2339: Property 'beacon_state' does not exist on type 'string | object'.
  Property 'beacon_state' does not exist on type 'string'.
src/lib.ts(248,23): error TS2532: Object is possibly 'undefined'.
src/lib.ts(248,32): error TS2339: Property 'shard_blocks' does not exist on type 'string | object'.
  Property 'shard_blocks' does not exist on type 'string'.
src/lib.ts(252,38): error TS2532: Object is possibly 'undefined'.
src/lib.ts(252,47): error TS2339: Property 'shard_pre_state' does not exist on type 'string | object'.
  Property 'shard_pre_state' does not exist on type 'string'.
src/lib.ts(253,39): error TS2532: Object is possibly 'undefined'.
src/lib.ts(253,48): error TS2339: Property 'shard_post_state' does not exist on type 'string | object'.
  Property 'shard_post_state' does not exist on type 'string'.
```

ethereumjs-config broken upstream :/

**Comment by jwasinger** [comment]
> ```
> > ethereumjs-config-build
> 
> + exec tsc -p ./tsconfig.prod.json
> src/lib.ts(247,19): error TS2532: Object is possibly 'undefined'.
> src/lib.ts(247,28): error TS2339: Property 'beacon_state' does not exist on type 'string | object'.
>   Property 'beacon_state' does not exist on type 'string'.
> src/lib.ts(248,23): error TS2532: Object is possibly 'undefined'.
> src/lib.ts(248,32): error TS2339: Property 'shard_blocks' does not exist on type 'string | object'.
>   Property 'shard_blocks' does not exist on type 'string'.
> src/lib.ts(252,38): error TS2532: Object is possibly 'undefined'.
> src/lib.ts(252,47): error TS2339: Property 'shard_pre_state' does not exist on type 'string | object'.
>   Property 'shard_pre_state' does not exist on type 'string'.
> src/lib.ts(253,39): error TS2532: Object is possibly 'undefined'.
> src/lib.ts(253,48): error TS2339: Property 'shard_post_state' does not exist on type 'string | object'.
>   Property 'shard_post_state' does not exist on type 'string'.
> ```
> 
> ethereumjs-config broken upstream :/

Actually a few of our `scout.ts-*` repos needed to be modified to work with Node 12.  I did that in https://github.com/ewasm-benchmarking/scout.ts/tree/bn128-pairing and https://github.com/ewasm-benchmarking/scout.ts/tree/secp-sig-verify .  Waiting on the bench run to finish.

**Comment by jwasinger** [comment]
> Will squash `RUN` statements after I verify that these changes work.

Tried this and it didn't save me much space at all.

**Pull request merged**